Receiver CRON network results dashboard

TomNUSDS opened the pull request:
Fixes https://github.com/CDCgov/prime-reportstream/issues/5679

## Test Steps:
1. In client, select Admin -> Receiver status dashboard
Verify:
 - [ ] Change start and end dates and verify data changes
 - [ ] Type in filters for Receiver and Results; verify data changes
 - [ ] Select Mixed Results from the Success dropdown
 - [ ] Click on Green time slice and Red time slice and verify Info Details dialog is displayed.
 
## Todo

Not part of this PR but future work.

- Refactor the one major rendering component into subcomponents to make it easier to test/maintain.
- Cache (memo) each row component so filtering is faster

## Changes
<img width="450" alt="image" src="https://user-images.githubusercontent.com/74203452/180578638-713f87c4-b409-4135-bb70-309cc06451be.png">

 - Updated the admin.yml API file to include new Admin-only API call.

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [x] **N/A** Are there licensing issues with any new dependencies introduced? 
- [x] Includes a summary of what a code reviewer should test/verify?
- [x] **N/A** Updated the release notes?
- [x] **N/A** Database changes are submitted as a separate PR?
- [x] **N/A** DevOps team has been notified if PR requires ops support?

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

TomNUSDS commented on frontend-react/src/components/Admin/AdminLastMileFailuresTable.tsx:
```diff
@@ -115,184 +115,178 @@ ${data.receiver}`;
-                            </th>
-                        </tr>
-                    </thead>
+        <section className="grid-container rs-container-unbounded">
```
This is just removing the <></> Turn off white-space diffs for this file

TomNUSDS replied:
Oh, you mean in a different PR? Sure!

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+                                ) {
+                                    return;
+                                }
+                                const clickKey =
```
[another nit] `parseInt("") === NaN; keys[NaN] === undefined` so, why not
```suggestion
                                const clickKey = evt.currentTarget?.dataset["keyoffset"] ? parseInt(evt.currentTarget.dataset["keyoffset"]) : "";
```
this is a dumb comment, but presenting an explicit code path that results in evaluating `parseInt("")` and then using that to index into an array makes me itchy

TomNUSDS replied:
Good feedback. Done

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+                            className={`slice ${sliceClassName}`}
+                            data-keyoffset={keyOffset - 1}
+                            onClick={(evt) => {
+                                if (
```
a very minor optimization, but we can slightly reduce complexity of this handler, since we know at render time whether or not it will be an early bail. something like
```
const clickHandler =  successForSlice.currentState === SuccessRate.UNDEFINED ? noop : actualClickHandler
```

TomNUSDS replied:
Excellent point. I'll try to incorporate it.

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+                ref={modalShowInfoRef}
+                id={"showSuccessDetails"}
+            >
+                {ModalInfoRender({ subData: currentDataForModal })}
```
why not treat this like a React component?

TomNUSDS replied:
Changing to 
```html
<ModalInfoRender subData={currentDataForModal} />
```

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+        return <>No Data Found</>;
+    }
+    // note: if we ever have the timeslots > cron job so there are multiple
+    // resuls per slot, then this needs to be expended to show more.
```
```suggestion
    // results per slot, then this needs to be expended to show more.
```

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+    const showDetailsModal = (subData: AdmConnStatusDataType[]) => {
+        if (subData.length) {
+            setCurrentDataForModal(subData);
+            modalShowInfoRef?.current?.toggleModal(undefined, true);
```
[note to self] test this out

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+            <section>
+                CRON job results that check if receivers are working.
+            </section>
+            <form autoComplete="off" className="grid-row margin-0">
```
I know you mentioned most of this not being reusable, but these date and string based filters could potentially be really useful in other views. Not important to split these out now, but wanted to plant the seed of an idea

doug-s-nava commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+                    >
+                        Results:
+                    </Label>
+                    <TextInput
```
it wasn't entirely clear to me what this filter was filtering on, also don't know how to make it more clear other than adding a tool tip? Is there a content / design review for this on the table?

TomNUSDS replied:
It's an admin-only page so there's no formal design or user testing since our internal O&O team are the only users. To date, the design team has been very busy with user-facing features, so their time is too valuable for the first pass n an internal-only dashboard. @drew-usds 

And you or an O&O team member are literally the user testing and I'm HAPPY to make any design changes

**Why it functions the way it does:** 
There is an error field that isn't showing in the table view and only shown in the details dialog because it's just too long. 

I played with having a single filter for both the visible data in the table (e.g. Org or Receiver) AND the buried error messages but it made the filter feel unpredictable. 

I can add hover help OR a tooltip or OR some better description text. Do you have a preference?

TomNUSDS replied:
Added `<Tooltip>` to give more UI discoverability

doug-s-nava reviewed (commented):
looks great (please don't be scared off by 25 comments)

TomNUSDS commented on frontend-react/src/pages/admin/AdminReceiverDashPage.tsx:
```diff
@@ -0,0 +1,34 @@
+
+export function AdminReceiverDashPage() {
+    return (
+        <NetworkErrorBoundary
```
@kevinhaube's thoughts on this.

It would be best if network errors were handled in component that generated them, but the way coinbase's RestHooks works is that the error handling happens OUTSIDE the component.

kevinhaube replied:
The short of it is NetworkErrorBoundary is from `rest-hooks` but picks up any network error. Think of it like a catch block for components. It'll catch whatever throws, and render whatever the ErrorBoundary component tells it to render. We could (and likely will) make our own and show different things based on error codes. Suspense is the equivalent of `if (loading) <Spinner />` as a higher-order component for _everything_ it wraps...so if any of those components wrapped in it fetch anything, until the last one loads its data, we'll have a spinner.

There's some helpful info in our Tech-Debt ticket about this!

TomNUSDS replied:
Yeah, I just followed the format of the outer page I saw in other files and didn't give much thought to it. I'll investigate what the right answer is.

TomNUSDS replied:
I removed the outer `<NetworkErrorBoundary>` in the latest update

TomNUSDS commented on frontend-react/src/resources/AdmConnStatusResource.ts:
```diff
@@ -0,0 +1,77 @@
+import AuthResource from "./AuthResource";
```
@kevinhaube 
Agreed, I started this PR _trying_ to use the new approach, but struggled to figure it out. Folks weren't available for questions (OOO), so I just started with what I knew would worked. The old network code approach is very concise. 

`AdmConnStatusResource.ts` without filters and type declarations is like 6 lines of actual code.

I'd be happy to refactor, but having a _complex_ example of how to use the new UseRequestConfig would be super useful. For example, how to we do get the `AdmConnStatusDataType` shape into the declaration?

kevinhaube replied:
Yeah, the new API work is also definitely _rough_ -- I think it makes more sense to implement it small-scale first with unique cases like Daily Data and Uploader, but there's lots of kinks to work out. I wish I was good enough to get it all packed in on the first attempt 😂

TomNUSDS commented on prime-router/docs/api/admin.yml:
```diff
@@ -1,4 +1,4 @@
-openapi: 3.0.2
+openapi: 3.1.0
```
Yeah. Bumping the version is problematic. Thanks for flagging!

I struggled to get both API calls documented correctly using 3.0.2 standard, but 3.1 was fine. The primary issue with going to 3.1 is tool compatibility. Many API tools don't fully support 3.1

I'll revisit how to fix this now. I'm taking on the task of fixing up the OpenAPI yml docs, so I'm going to have to address it somehow. 

I suspect I'm going to have to have two files. One for `/adm/getsendfailures` and another for `/adm/listreceiversconnstatus`, however, it's just a bunch of copy/pasting.

TomNUSDS replied:
I completely reworked this file to be compatible with 3.0.2. Marking resolved.

TomNUSDS commented on prime-router/src/main/kotlin/azure/AdminApiFunctions.kt:
```diff
@@ -59,18 +60,56 @@ class AdminApiFunctions(
+    @FunctionName("listreceiversconnstatus")
+    fun listReceiversConnStatus(
+        @HttpTrigger(
+            name = "listreceiversconnstatus",
```
Yeah, I suspect we may have standardized on the wrong casing.

There's an API working group meeting next week. We should add this topic to the discussion.

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+                    </Dropdown>
+                </div>
+            </form>
+            {MainRender({
```
Umm... good point. Same functionality I think, but it's not very React of me.

Changing to this:

```
            <MainRender
                data={data}
                filterRowStatus={filterRowSuccessState}
                datesRange={[
                    new Date(startDate),
                    endDate ? new Date(endDate) : new Date(),
                ]}
                onDetailsClick={showDetailsModal}
            />
```

TomNUSDS commented on frontend-react/src/styles/_reportstream.scss:
```diff
@@ -1,183 +1,176 @@
 // USWDS identifier
 
 .prime-identifier {
```
Agreed. This doesn't need to be in this PR. Reverting whitespace changes.

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+ * @param d {Date}
+ * @return {string}
+ */
+const defaultStartDateIso = (d = new Date()) => {
```
I'd considered putting them there, but these seemed too specific. 

If you're ok with it, I'd wait until we need these functions in more than one place?

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+    for (let item of dataIn) {
+        const key = `${item.organizationName}|${item.receiverName}\t${item.connectionCheckStartedAt}`;
+        data[key] = item;
+        if (lastkey > key) {
```
Yeah. This keys-determine-sort-order is a concept take over from [key-value databases](https://en.wikipedia.org/wiki/Key%E2%80%93value_database).

The `MainRender` does test that the data resorts correctly for `makeDictionary`. (The data in `receiverConnectionCheckResultId: 2396` is out of order in `DATA` for the test.)

But in the end, the code probably doesn't really need keys since keys:value pairs aren't actually sorted by keys in javascript like they are in other languages. (I actually forgot this detail when I planned the code. Too many languages crammed into my head.)

That said, sorting strings is crazy FAST. And the reason why ISO dates are laid out the way they are is specific to descending resolution Year > month > day > hour > min > sec WITH leading zeros so they sort correctly as strings.

Sort order for ISO string > ISO string is the same as ISO Date > ISO Date.

TomNUSDS replied:
That said, simplifying this is a pretty complex refactor. I'd like to save it for another round if that's ok.

I'd like to tackle speeding up filtering in a refactor as well.

But "perfect can be the enemy of better", so I'd like to get this checked in, then iterate on it.

I put it in #6235 to keep track of it.

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+                }
+
+                // render slots for day
+                {
```
Yeah, these brackets are intended to be future component boundaries. They can be removed, but they were helpful visual indicators as I was coding.

doug-s-nava replied:
yeah, I'd either remove them or add a comment about the component boundaries. The brackets to me (maybe just to me) read as possible JSX syntax which is a little confusing

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+// key/value pair
+type DataDictionary = Record<string, AdmConnStatusDataType>;
+// prop drilled down to bottom most components
+// type dateClickHandler = (id: string) => void;
```
good catch. removed

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+        return <div>No Data</div>;
+    }
+
+    const onClick = (clickedKey: string) => {
```
done

TomNUSDS commented on frontend-react/src/components/Admin/AdminReceiverDashboard.tsx:
```diff
@@ -0,0 +1,775 @@
+        do {
+            yield [
+                this.current,
+                dateAddHours(this.current, this.skipHours),
```
This iterator ended up being a little slow when I profiled the code, so the function has been optimized to only call dateAddHours once per iteration.

TomNUSDS commented:
@doug-s-nava Thanks for the detailed review. Super helpful. I think I've addressed most of your feedback and/or made another bug to improve if it seemed like to big of a refactor. Can you look it over and see if I missed anything?

doug-s-nava commented on frontend-react/src/App.tsx:
```diff
@@ -225,6 +226,11 @@ const App = () => {
                                     component={AdminLastMileFailures}
                                 />
+                                <AuthorizedRoute
+                                    path="/admin/senddash"
```
another "case" where there are two conventions going. Routes seem to be using both kebab case and flat case. I'm pretty hardcore in favor of kebab case here, but we can wait and discuss tomorrow at eng sync?

TomNUSDS replied:
So for renaming the url path on our site, I'm happy to change it to whatever we decide is should be. 
`/admin/send-dash` (kebab)
`/admin/send_dash` (snake)
`/admin/sendDash` (camel)

If it's related to the API call, suspect it will be punted to the next API working group meeting.

I've started a list of API inconsistencies here #6243, you should add your api naming convention preference to it. I suspect we're in the the situation we're in now because nobody had a strong enough opinion ... and naming is hard.

doug-s-nava reviewed (commented):
This is very close. Thanks for bearing with me and making all these changes Tom. The only thing I still want to make sure about before approving is the two ErrorBoundaries/Suspenses on the page.

TomNUSDS commented:
@doug-s-nava I think I've addressed everything including remove the extra code scoping brackets. (I used the IDK hint `// <editor-fold desc="Time slices for a given day render">` .. which webstorm users but I'm not sure about vscode.)

Think we're close. Thanks again for all the time spent reviewing!

doug-s-nava reviewed (approved):
✅ ✅ ✅ ✅ ✅ 

on the error boundary question, I'm still not clear on why we need two, but we do have a card scheduled to refactor that behavior across the site in general, so if there is an improvement to make there it will likely be made at that point.

nice work, appreciate you weathering my comments

TomNUSDS commented:
@doug-s-nava I really do appreciate a detailed review and the time, focus it requires. Thank-you for being so thorough.
